### PR TITLE
Secret storage on device

### PR DESCRIPTION
**Pull request by AOsnes**

### CODE REVIEW

File: client/src/encryption.rs [hunk 1]
```diff
@@ -69,6 +70,28 @@ pub(crate) mod test {
         InMemSignalProtocolStore::new(p, reg).unwrap()
     }
 
+    pub fn signal_bundle_to_our_bundle(bundle: PreKeyBundle) -> KeyBundleContent {
+        KeyBundleContent::new(
+            bundle.registration_id().unwrap(),
+            bundle.device_id().unwrap(),
+            Some((
+                bundle.pre_key_id().unwrap().unwrap(),
```
Comment (Mikkel-Boje) [inline client/src/encryption.rs:78]: No unwrap

File: client/src/key_management/bundle.rs [hunk 2]
```diff
@@ -53,4 +55,130 @@ impl KeyBundleContent {
             None => Ok(pre_key_bundle),
         }
     }
+
+    pub fn serialize(&self) -> PrimitiveKeyBundleContent {
+        PrimitiveKeyBundleContent::new(
+            self.registration_id,
+            self.device_id.into(),
+            Some(self.onetime_public_key.unwrap().0.into()),
```
Comment (Mikkel-Boje) [inline client/src/key_management/bundle.rs:63]: No unwrap

**Review by Mikkel-Boje — DISMISSED** [review]

File: client/src/encryption.rs [hunk 3]
```diff
@@ -69,6 +70,28 @@ pub(crate) mod test {
         InMemSignalProtocolStore::new(p, reg).unwrap()
     }
 
+    pub fn signal_bundle_to_our_bundle(bundle: PreKeyBundle) -> KeyBundleContent {
```
Comment (UnluckyBird) [inline client/src/encryption.rs:73]: Maybe implement `impl From<PreKeyBundle> for KeyBundleContent` instead?

File: client/src/key_management/bundle.rs [hunk 4]
```diff
@@ -53,4 +55,230 @@ impl KeyBundleContent {
...
+                || bundle.onetime_public_key.is_none()
+                || bundle
+                    .onetime_public_key_id
+                    .expect("One time pre key id is somehow None after a none check - kill me")
+                    == 0)
+            {
+                None
+            } else {
+                Some((
+                    bundle.onetime_public_key_id.expect("").into(),
```
Comment (UnluckyBird) [inline client/src/key_management/bundle.rs:101]: Use unwrap instead of expect if no custom error message is chosen anyway

**Review by UnluckyBird — APPROVED** [review]
Should probably replace some of the unwraps for question marks to make the code less panicky

File: client/src/encryption.rs (hunk 3 above)
Comment (AOsnes) [reply client/src/encryption.rs:73]: cba når det er en test

File: client/src/key_management/bundle.rs (hunk 4 above)
Comment (AOsnes) [reply client/src/key_management/bundle.rs:101]: Mikkel er en unwrap hater så nu er der expect i stedet for

**Review by Laursen79 — APPROVED** [review]

**Review by Mikkel-Boje — APPROVED** [review]
:+1:

**Pull request merged**